src/main/java/net/fabricmc/loader/minecraft/McVersionLookup.java

```diff
@@ -0,0 +1,555 @@
+			if (Files.isRegularFile(file = fs.getPath("net/minecraft/realms/RealmsBridge.class"))
+					&& (ret = fromAnalyzer(Files.newInputStream(file), new MethodConstantRetVisitor("getVersionString"))) != null) {
+				return ret;
+			}
+
+			// version-like String constant used in MinecraftServer.run or another MinecraftServer method
+			if (Files.isRegularFile(file = fs.getPath("net/minecraft/server/MinecraftServer.class"))
+					&& (ret = fromAnalyzer(Files.newInputStream(file), new MethodConstantVisitor("run"))) != null) {
+				return ret;
+			}
+
+			// version-like constant return value of a Minecraft method (obfuscated/unknown name)
+			if (Files.isRegularFile(file = fs.getPath("net/minecraft/client/Minecraft.class"))
+					&& (ret = fromAnalyzer(Files.newInputStream(file), new MethodConstantRetVisitor(null))) != null) {
+				return ret;
+			}
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
+
+		return fromFileName(gameJar.getFileName().toString());
+	}
+
+	private static McVersion fromVersionJson(InputStream is) {
+		try (JsonReader reader = new JsonReader(new InputStreamReader(is, StandardCharsets.UTF_8))) {
+			String name = null;
+			String release = null;
+
+			reader.beginObject();
+
+			while (reader.hasNext()) {
+				switch (reader.nextName()) {
+				case "name": name = reader.nextString(); break;
+				case "release_target": release = reader.nextString(); break;
+				default: reader.skipValue();
+				}
+			}
+
+			reader.endObject();
+
+			if (name != null && release != null) return new McVersion(name, release);
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
+
+		return null;
+	}
+
+	private static <T extends ClassVisitor & Analyzer> McVersion fromAnalyzer(InputStream is, T analyzer) {
+		try {
+			ClassReader cr = new ClassReader(is);
+			cr.accept(analyzer, ClassReader.SKIP_FRAMES | ClassReader.SKIP_DEBUG);
+			String result = analyzer.getResult();
+
+			if (result != null) {
+				return new McVersion(result, getRelease(result));
+			}
+		} catch (IOException e) {
+			e.printStackTrace();
+		} finally {
+			try {
+				is.close();
+			} catch (IOException e) { }
+		}
+
+		return null;
+	}
+
+	private static McVersion fromFileName(String name) {
+		// strip extension
+		int pos = name.lastIndexOf('.');
+		if (pos > 0) name = name.substring(0, pos);
+
+		return new McVersion(name, getRelease(name));
+	}
+
+	private static String getRelease(String version) {
+		if (RELEASE_PATTERN.matcher(version).matches()) return version;
+
+		assert isProbableVersion(version);
+
+		int pos = version.indexOf("-pre");
+		if (pos >= 0) return version.substring(0, pos);
+
+		pos = version.indexOf(" Pre-Release ");
+		if (pos >= 0) return version.substring(0, pos);
+
+		Matcher matcher = SNAPSHOT_PATTERN.matcher(version);
+
+		if (matcher.matches()) {
+			int year = Integer.parseInt(matcher.group(1));
+			int week = Integer.parseInt(matcher.group(2));
+
+			if (year == 19 && week == 34) {
```

Comment:
Might want to do `week >= 34`, I know we dont have an upper bound on this just yet.

In newer versions you could use `release_target` from the version.json file in the jar, this should remove the need to keep this list maintained as time passes.

Reply:
It already uses the json if available, the implementation is conservative to avoid associating a release incorrectly, which would bypass the planned MC version deps unintentionally.

Reply:
Ah yeah, didnt see that bit, thanks for clarifying.